File: data/derived_cot/rq3_prompting/results/translation/r1/java_to_py/code_output/VectorUtil.py
import math
from typing import List, Dict

class VectorUtil:

    @staticmethod
    def similarity(vector1: List[float], vector2: List[float]) -> float:
        dotProduct = 0.0
        norm1 = 0.0
        norm2 = 0.0
        for i in range(len(vector1)):
            dotProduct += vector1[i] * vector2[i]
            norm1 += vector1[i] ** 2
            norm2 += vector2[i] ** 2
        denominator = math.sqrt(norm1) * math.sqrt(norm2)
        return 0.0 if denominator == 0 else dotProduct / denominator

    @staticmethod
    def cosineSimilarities(vector1: List[float], vectorsAll: List[List[float]]) -> List[float]:
        similarities = []
        norm1 = math.sqrt(VectorUtil._dotProduct(vector1, vector1))
        if norm1 == 0:
            for _ in vectorsAll:
                similarities.append(0.0)
            return similarities
        for vector2 in vectorsAll:
            norm2 = math.sqrt(VectorUtil._dotProduct(vector2, vector2))
            if norm2 == 0:
                similarities.append(0.0)
            else:
                sim = VectorUtil._dotProduct(vector1, vector2) / (norm1 * norm2)
                similarities.append(sim)
        return similarities

    @staticmethod
    def nSimilarity(vectorList1: List[List[float]], vectorList2: List[List[float]]) -> float:
        if not vectorList1 or not vectorList2:
            raise ValueError("At least one of the passed lists is empty.")
        avgVector1 = VectorUtil._averageVector(vectorList1)
        avgVector2 = VectorUtil._averageVector(vectorList2)
        return VectorUtil.similarity(avgVector1, avgVector2)

    @staticmethod
    def computeIdfWeightDict(totalNum: int, numberDict: Dict[str, float]) -> Dict[str, float]:
        result = {}
        for key, count in numberDict.items():
            weight = math.log((totalNum + 1) / (count + 1))
            result[key] = weight
        return result

    @staticmethod
    def _dotProduct(vector1: List[float], vector2: List[float]) -> float:
        result = 0.0
        for i in range(len(vector1)):
            result += vector1[i] * vector2[i]
        return result

    @staticmethod
    def _averageVector(vectors: List[List[float]]) -> List[float]:
        avg = [0.0] * len(vectors[0])
        for vec in vectors:
            for i in range(len(vec)):
                avg[i] += vec[i]
        for i in range(len(avg)):
            avg[i] /= len(vectors)
        return avg

File: data/derived_cot/rq2_eval/translation/qwen/_converted_code/java_to_cpp/VectorUtil.cpp
#include <vector>
#include <map>
#include <cmath>
#include <stdexcept>
#include <utility>

class VectorUtil {

public:
    static double similarity(const std::vector<double>& vector1, const std::vector<double>& vector2) {
        if (vector1.size() != vector2.size()) {
            throw std::invalid_argument("Vectors must be of the same length");
        }

        double dotProduct = 0.0;
        double norm1 = 0.0;
        double norm2 = 0.0;

        for (size_t i = 0; i < vector1.size(); ++i) {
            dotProduct += vector1[i] * vector2[i];
            norm1 += vector1[i] * vector1[i];
            norm2 += vector2[i] * vector2[i];
        }

        double denominator = std::sqrt(norm1) * std::sqrt(norm2);
        return denominator == 0.0 ? 0.0 : dotProduct / denominator;
    }

    static std::vector<double> cosineSimilarities(const std::vector<double>& vector1, const std::vector<std::vector<double>>& vectorsAll) {
        std::vector<double> similarities;

        // Compute norm of vector1
        double norm1 = std::sqrt(dotProduct(vector1, vector1));
        if (norm1 == 0.0) {
            // If vector1 is zero, all similarities are zero
            for (const auto& vec : vectorsAll) {
                similarities.push_back(0.0);
            }
            return similarities;
        }

        for (const auto& vector2 : vectorsAll) {
            // Check if vector2 is valid
            if (vector1.size() != vector2.size()) {
                throw std::invalid_argument("Vectors must be of the same length");
            }

            double norm2 = std::sqrt(dotProduct(vector2, vector2));
            if (norm2 == 0.0) {
                similarities.push_back(0.0);
            } else {
                similarities.push_back(dotProduct(vector1, vector2) / (norm1 * norm2));
            }
        }

        return similarities;
    }

    static double n_similarity(const std::vector<std::vector<double>>& vectorList1, const std::vector<std::vector<double>>& vectorList2) {
        if (vectorList1.empty() || vectorList2.empty()) {
            throw std::invalid_argument("At least one of the passed lists is empty.");
        }

        auto avgVector1 = averageVector(vectorList1);
        auto avgVector2 = averageVector(vectorList2);
        return similarity(avgVector1, avgVector2);
    }

    static std::map<std::string, double> computeIdfWeightDict(int totalNum, const std::map<std::string, double>& numberDict) {
        std::map<std::string, double> result;
        for (const auto& entry : numberDict) {
            double count = entry.second;
            double weight = std::log(static_cast<double>(totalNum + 1) / (count + 1));
            result[entry.first] = weight;
        }
        return result;
    }

private:
    static double dotProduct(const std::vector<double>& vector1, const std::vector<double>& vector2) {
        if (vector1.size() != vector2.size()) {
            throw std::invalid_argument("Vectors must be of the same length");
        }

        double result = 0.0;
        for (size_t i = 0; i < vector1.size(); ++i) {
            result += vector1[i] * vector2[i];
        }
        return result;
    }

    static std::vector<double> averageVector(const std::vector<std::vector<double>>& vectors) {
        if (vectors.empty()) {
            throw std::invalid_argument("Cannot compute average of empty vector list");
        }

        size_t dimension = vectors[0].size();
        std::vector<double> avgVector(dimension, 0.0);

        for (const auto& vec : vectors) {
            if (vec.size() != dimension) {
                throw std::invalid_argument("Vectors in the list must have the same dimension");
            }
            for (size_t i = 0; i < dimension; ++i) {
                avgVector[i] += vec[i];
            }
        }

        double numVectors = vectors.size();
        for (size_t i = 0; i < dimension; ++i) {
            avgVector[i] /= numVectors;
        }

        return avgVector;
    }
};